mantidimaging/core/io/filenames.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (C) 2022 ISIS Rutherford Appleton Laboratory UKRI
+# SPDX - License - Identifier: GPL-3.0-or-later
+
+from pathlib import Path
+import re
+from typing import List, Iterator, Optional
+
+
+class FilenamePattern:
+    """
+    Representation of a filename pattern to handle stacks of related images
+
+    Handles patterns like "aaaa_####.bbb" where # are digits
+    """
+    PATTERN_p = r'^([a-zA-Z0-9_]+?)'
+    PATTERN_d = r'([0-9]+)'
+    PATTERN_s = r'(\.[a-zA-Z_]+)$'
+    PATTERN = re.compile(PATTERN_p + PATTERN_d + PATTERN_s)
+
+    def __init__(self, prefix: str, digit_count: int, suffix: str):
+        self.prefix = prefix
+        self.digit_count = digit_count
+        self.suffix = suffix
+
+        if digit_count == 0:
+            self.re_pattern = re.compile("^" + re.escape(prefix) + re.escape(suffix) + "$")
+        else:
+            # Note: allow extra leading digits, for data sets that go 001 ... 998, 999, 1000, 1001
+            self.re_pattern = re.compile("^" + re.escape(prefix) + "([1-9]?[0-9]{" + str(digit_count) + "})" +
+                                         re.escape(suffix) + "$")
+
+        self.re_pattern_metadata = re.compile("^" + re.escape(prefix.rstrip("_ ")) + ".json")
+        self.template = prefix + "{:0" + str(digit_count) + "d}" + suffix
+
+    @classmethod
+    def from_name(cls, filename: str) -> "FilenamePattern":
+        result = FilenamePattern.PATTERN.search(filename)
+
+        if result is None:
+            if "." in filename:
+                name, _, ext = filename.rpartition(".")
+                return FilenamePattern(name, 0, "." + ext)
+            else:
+                return FilenamePattern(filename, 0, "")
+
+        prefix = result.group(1)
+        digits = result.group(2)
+        ext = result.group(3)
+
+        return FilenamePattern(prefix, len(digits), ext)
+
+    def generate(self, index: int) -> str:
+        if self.digit_count == 0:
+            return self.prefix + self.suffix
+        return self.template.format(index)
+
+    def match(self, filename: str) -> bool:
+        return self.re_pattern.match(filename) is not None
+
+    def get_value(self, filename: str) -> int:
```

Comment:
Would this make sense to rename to `get_index`? As I was reading through it felt like it might be more consistent with the naming used elsewhere.